File: scripts/fail2ban-whitelist.py
#!/usr/bin/env python3
"""Idempotently add ovpn_mgr's OpenVPN IPv4 pool to Fail2Ban DEFAULT ignoreip."""
import ipaddress
import os
import re
import shutil
import sys
import tempfile
from datetime import datetime

if len(sys.argv) != 3:
    print(f"Usage: {sys.argv[0]} OPENVPN_CONF JAIL_LOCAL", file=sys.stderr)
    sys.exit(2)

conf_path, jail_path = sys.argv[1:]
try:
    conf_text = open(conf_path, encoding="utf-8").read()
except OSError as exc:
    print(f"Cannot read OpenVPN config {conf_path}: {exc}", file=sys.stderr)
    sys.exit(1)

network = None
for line in conf_text.splitlines():
    match = re.match(r"^\s*server\s+(\S+)\s+(\S+)", line)
    if match:
        try:
            network = str(ipaddress.IPv4Network(f"{match.group(1)}/{match.group(2)}", strict=False))
            break
        except (ipaddress.AddressValueError, ipaddress.NetmaskValueError):
            continue
if not network:
    print("Could not determine a valid IPv4 OpenVPN 'server <network> <netmask>' pool; Fail2Ban whitelist unchanged.", file=sys.stderr)
    sys.exit(1)

try:
    original = open(jail_path, encoding="utf-8").read() if os.path.exists(jail_path) else ""
except OSError as exc:
    print(f"Cannot read {jail_path}: {exc}", file=sys.stderr)
    sys.exit(1)

lines = original.splitlines()
section_start = section_end = None
for idx, line in enumerate(lines):
    if re.match(r"^\s*\[\s*DEFAULT\s*\]\s*(?:[;#].*)?$", line, re.I):
        section_start = idx
        section_end = len(lines)
        for j in range(idx + 1, len(lines)):
            if re.match(r"^\s*\[", lines[j]):
                section_end = j
                break
        break

if section_start is None:
    if lines and lines[-1].strip():
        lines.append("")
    lines.extend(["[DEFAULT]", f"ignoreip = 127.0.0.1/8 ::1 {network}"])
else:
    ignore_idx = None
    for idx in range(section_start + 1, section_end):
        if re.match(r"^\s*ignoreip\s*=", lines[idx], re.I):
            ignore_idx = idx
            break
    if ignore_idx is None:
        lines.insert(section_end, f"ignoreip = 127.0.0.1/8 ::1 {network}")
    else:
        # Keep any inline comment and all pre-existing ignore entries intact.
        before_comment, sep, comment = lines[ignore_idx].partition("#")
        if not sep:
            before_comment, sep, comment = before_comment.partition(";")
        match = re.match(r"^(\s*ignoreip\s*=\s*)(.*)$", before_comment, re.I)
        value = match.group(2).strip() if match else ""
        entries = value.split()
        if network not in entries:
            lines[ignore_idx] = match.group(1) + (value + " " if value else "") + network
            if sep:
                lines[ignore_idx] += " " + sep + comment

updated = "\n".join(lines).rstrip() + "\n"
if updated == original:
    print(f"Fail2Ban already includes OpenVPN pool {network} in {jail_path}.")
    print(network)
    sys.exit(0)

os.makedirs(os.path.dirname(jail_path) or ".", exist_ok=True)
if os.path.exists(jail_path):
    backup = jail_path + ".ovpn_mgr.bak"
    if not os.path.exists(backup):
        shutil.copy2(jail_path, backup)
    mode = os.stat(jail_path).st_mode & 0o777
else:
    mode = 0o644
fd, tmp = tempfile.mkstemp(prefix=".jail.local.", dir=os.path.dirname(jail_path) or ".", text=True)
try:
    with os.fdopen(fd, "w", encoding="utf-8") as handle:
        handle.write(updated)
    os.chmod(tmp, mode)
    os.replace(tmp, jail_path)
except Exception:
    try:
        os.unlink(tmp)
    except OSError:
        pass
    raise
print(f"Added OpenVPN pool {network} to {jail_path} [DEFAULT] ignoreip.")
print(network)

File: scripts/fail2ban-whitelist.php
<?php
/**
 * Idempotently add ovpn_mgr's OpenVPN IPv4 pool to Fail2Ban [DEFAULT] ignoreip.
 * Usage: php fail2ban-whitelist.php OPENVPN_CONF JAIL_LOCAL
 * Runs as root from setup-root.sh. PHP is used (not Python) because it is
 * always present on FreePBX, unlike python3 on some FreePBX 16 installs.
 */
if (PHP_SAPI !== 'cli') {
    fwrite(STDERR, "This helper must run from the CLI.\n");
    exit(2);
}
if ($argc !== 3) {
    fwrite(STDERR, "Usage: {$argv[0]} OPENVPN_CONF JAIL_LOCAL\n");
    exit(2);
}
[$confPath, $jailPath] = [$argv[1], $argv[2]];

$confText = is_readable($confPath) ? file_get_contents($confPath) : false;
if ($confText === false) {
    fwrite(STDERR, "Cannot read OpenVPN config {$confPath}\n");
    exit(1);
}

// Find the first valid "server <network> <netmask>" line.
$network = null;
foreach (preg_split('/\r\n|\r|\n/', $confText) as $line) {
    if (!preg_match('/^\s*server\s+(\S+)\s+(\S+)/', $line, $m)) {
        continue;
    }
    $ipLong = filter_var($m[1], FILTER_VALIDATE_IP, FILTER_FLAG_IPV4) ? ip2long($m[1]) : false;
    $maskLong = filter_var($m[2], FILTER_VALIDATE_IP, FILTER_FLAG_IPV4) ? ip2long($m[2]) : false;
    if ($ipLong === false || $maskLong === false) {
        continue;
    }
    $maskBin = sprintf('%032b', $maskLong & 0xffffffff);
    if (strpos($maskBin, '01') !== false) { // non-contiguous netmask
        continue;
    }
    $network = long2ip(($ipLong & $maskLong) & 0xffffffff) . '/' . substr_count($maskBin, '1');
    break;
}
if ($network === null) {
    fwrite(STDERR, "Could not determine a valid IPv4 OpenVPN 'server <network> <netmask>' pool; Fail2Ban whitelist unchanged.\n");
    exit(1);
}

$original = '';
if (file_exists($jailPath)) {
    $read = is_readable($jailPath) ? file_get_contents($jailPath) : false;
    if ($read === false) {
        fwrite(STDERR, "Cannot read {$jailPath}\n");
        exit(1);
    }
    $original = $read;
}

$lines = ($original === '') ? [] : preg_split('/\r\n|\r|\n/', $original);
if (!empty($lines) && end($lines) === '') {
    array_pop($lines);
}

$sectionStart = $sectionEnd = null;
foreach ($lines as $idx => $line) {
    if (preg_match('/^\s*\[\s*DEFAULT\s*\]\s*(?:[;#].*)?$/i', $line)) {
        $sectionStart = $idx;
        $sectionEnd = count($lines);
        for ($j = $idx + 1; $j < count($lines); $j++) {
            if (preg_match('/^\s*\[/', $lines[$j])) {
                $sectionEnd = $j;
                break;
            }
        }
        break;
    }
}

if ($sectionStart === null) {
    if (!empty($lines) && trim(end($lines)) !== '') {
        $lines[] = '';
    }
    $lines[] = '[DEFAULT]';
    $lines[] = "ignoreip = 127.0.0.1/8 ::1 {$network}";
} else {
    $ignoreIdx = null;
    for ($idx = $sectionStart + 1; $idx < $sectionEnd; $idx++) {
        if (preg_match('/^\s*ignoreip\s*=/i', $lines[$idx])) {
            $ignoreIdx = $idx;
            break;
        }
    }
    if ($ignoreIdx === null) {
        array_splice($lines, $sectionEnd, 0, ["ignoreip = 127.0.0.1/8 ::1 {$network}"]);
    } else {
        // Keep any inline comment and all pre-existing ignore entries intact.
        $full = $lines[$ignoreIdx];
        $sep = '';
        $comment = '';
        $before = $full;
        $pos = strpos($full, '#');
        if ($pos === false) {
            $pos = strpos($full, ';');
        }
        if ($pos !== false) {
            $before = substr($full, 0, $pos);
            $sep = $full[$pos];
            $comment = substr($full, $pos + 1);
        }
        if (preg_match('/^(\s*ignoreip\s*=\s*)(.*)$/i', $before, $mm)) {
            $value = trim($mm[2]);
            $entries = $value === '' ? [] : preg_split('/\s+/', $value);
            if (!in_array($network, $entries, true)) {
                $lines[$ignoreIdx] = $mm[1] . ($value !== '' ? $value . ' ' : '') . $network;
                if ($sep !== '') {
                    $lines[$ignoreIdx] .= ' ' . $sep . $comment;
                }
            }
        }
    }
}

$updated = rtrim(implode("\n", $lines)) . "\n";
if ($updated === $original) {
    echo "Fail2Ban already includes OpenVPN pool {$network} in {$jailPath}.\n";
    echo $network . "\n";
    exit(0);
}

$dir = dirname($jailPath) ?: '.';
if (!is_dir($dir) && !mkdir($dir, 0755, true)) {
    fwrite(STDERR, "Cannot create {$dir}\n");
    exit(1);
}
if (file_exists($jailPath)) {
    $backup = $jailPath . '.ovpn_mgr.bak';
    if (!file_exists($backup)) {
        copy($jailPath, $backup);
    }
    $mode = fileperms($jailPath) & 0777;
} else {
    $mode = 0644;
}
$tmp = tempnam($dir, '.jail.local.');
if ($tmp === false || file_put_contents($tmp, $updated) === false) {
    if ($tmp !== false) { @unlink($tmp); }
    fwrite(STDERR, "Cannot write temporary file in {$dir}\n");
    exit(1);
}
chmod($tmp, $mode);
if (!rename($tmp, $jailPath)) {
    @unlink($tmp);
    fwrite(STDERR, "Cannot replace {$jailPath}\n");
    exit(1);
}
echo "Added OpenVPN pool {$network} to {$jailPath} [DEFAULT] ignoreip.\n";
echo $network . "\n";
